### PR TITLE
Draft: unix/trace: Add new variant to enable settrace.

### PR DESCRIPTION
**Pull request by andrewleech**

### CODE REVIEW

File: ports/unix/variants/trace/mpconfigvariant.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+// This config enables almost all possible features such that it can be used
+// for coverage testing.
```
Comment (jimmo) [inline ports/unix/variants/trace/mpconfigvariant.h:28]: Comment needs updating.

File: ports/unix/variants/trace/mpconfigvariant.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+#define MICROPY_CONFIG_ROM_LEVEL       (MICROPY_CONFIG_ROM_LEVEL_EVERYTHING)
+
+// Disable some features that come enabled by default with the feature level.
```
Comment (jimmo) [inline ports/unix/variants/trace/mpconfigvariant.h:32]: Why do these need to be disabled?

File: ports/unix/variants/trace/mpconfigvariant.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+#define MICROPY_PY_BUILTINS_EXECFILE            (0)
+#define MICROPY_PY_SYS_STDIO_BUFFER             (0)
+#define MICROPY_PY_USELECT                      (0)
+
+// Enable additional features.
```
Comment (jimmo) [inline ports/unix/variants/trace/mpconfigvariant.h:37]: Could this just include coverage/mpconfigvariant.h ?

File: ports/unix/variants/trace/mpconfigvariant.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+#define MICROPY_PY_URE_DEBUG           (1)
+#define MICROPY_PY_URE_MATCH_GROUPS    (1)
+#define MICROPY_PY_URE_MATCH_SPAN_START_END (1)
+#define MICROPY_PY_COLLECTIONS_NAMEDTUPLE__ASDICT (1)
+#define MICROPY_PY_UCRYPTOLIB          (1)
+#define MICROPY_PY_UCRYPTOLIB_CTR      (1)
+#define MICROPY_PY_MICROPYTHON_HEAP_LOCKED (1)
+
+
+#define MICROPY_PY_SYS_SETTRACE (1)
```
Comment (jimmo) [inline ports/unix/variants/trace/mpconfigvariant.h:61]: I wonder if this is better achieved by making a make variable that turns this on (and then works on all variants).

i.e. in unix/Makefile, if `MICROPY_PY_SYS_SETTRACE` is set, then add `CFLAGS=-DMICROPY_PY_SYS_SETTRACE` etc (and set DEBUG).

(Also newline end of file)

File: ports/unix/variants/trace/mpconfigvariant.mk [hunk 5]
```diff
@@ -0,0 +1,13 @@
+PROG ?= micropython-trace
+
+# Disable optimisations and enable assert() on trace builds.
+DEBUG ?= 1
+
+CFLAGS += \
+	-Wformat -Wmissing-declarations -Wmissing-prototypes \
```
Comment (jimmo) [inline ports/unix/variants/trace/mpconfigvariant.mk:7]: Are these necessary for settrace?

**Review by jimmo — COMMENTED** [review]
Thanks @andrewleech definitely useful to have an easy way to make a settrace build! Perhaps it doesn't need to be a variant though?

**Comment by dpgeorge** [comment]
`MICROPY_PY_SYS_SETTRACE` is enabled on the dev variant, so not sure what this adds?

**Comment by andrewleech** [comment]
For context, @mattytrentini, @HowManyOliversAreThere and I started working on `pdb` last night and made some really solid progress, this PR was initially just my easy way of sharing a method to build with trace turned on. It does need some cleanup certianly, but may also morph into a `pdb` variant with the debug support manifest'ed in - though `pdb` and deps belong in micropython-lib really.

**Comment by andrewleech** [comment]
> `MICROPY_PY_SYS_SETTRACE` is enabled on the dev variant, so not sure what this adds?

Ah.... we missed that :-D thanks. The second commit here will want to get split out into its own PR, but that will happen along with a bunch of other follow up `pdb` work!

**Pull request closed without merging**